actioncable/lib/action_cable/connection/subscriptions.rb

```diff
@@ -63,25 +65,39 @@ def unsubscribe_from_all
       private
         delegate :logger, to: :connection
 
-        def normalize_identifier(identifier)
-          identifier = ActiveSupport::JSON.encode(identifier) if identifier.is_a?(Hash)
-          identifier
+        def find(data)
+          if subscription = subscriptions[data['identifier']]
+            subscription
+          else
+            raise "Unable to find subscription with identifier: #{data['identifier']}"
+          end
         end
 
         # If `data` is a Hash, this means that the original JSON
         # sent by the client had no backslashes in it, and does
         # not need to be decoded again.
-        def decode_hash(data)
-          data = ActiveSupport::JSON.decode(data) unless data.is_a?(Hash)
-          data.with_indifferent_access
+        def normalize_data(data)
+          data = json_to_hash(data)
+
+          # Normalize the subscription's identifier
+          data['identifier'] = hash_to_json(data['identifier']) if data['identifier']
+
+          # Normalize the connection's message data
+          data['data'] = json_to_hash(data['data']) if data['data']
+
+          data
         end
 
-        def find(data)
-          if subscription = subscriptions[normalize_identifier(data['identifier'])]
-            subscription
-          else
-            raise "Unable to find subscription with identifier: #{data['identifier']}"
-          end
+        # Present the argument as a Hash
+        def json_to_hash(data)
+          data = ActiveSupport::JSON.decode(data).with_indifferent_access if !data.is_a?(Hash)
+          data
```

Comment:
Does different things than the old `decode_hash` method by not always calling `with_indifferent_access`. Think that breaks the contract.